### PR TITLE
[internal/otelarrow] Fix test flake (for 34719)

### LINKED ISSUE #34719: [internal/otelarrow] Flaky test disabled: TestIntegrationMemoryLimited
### Component(s)

internal/otelarrow

### Describe the issue you're reporting

Hit on #34358 see https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/10203882356/job/28231140032?pr=34358#step:6:518

```terminal
=== FAIL: test TestIntegrationSelfTracing (11.03s)
    e2e_test.go:369: 
        	Error Trace:	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:369
        	            				D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:205
        	            				C:/hostedtoolcache/windows/go/1.21.12/x64/src/runtime/asm_amd64.s:1650
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = send wait: context deadline exceeded
        	Test:       	TestIntegrationSelfTracing
    e2e_test.go:369: 
        	Error Trace:	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:369
        	            				D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:205
        	            				C:/hostedtoolcache/windows/go/1.21.12/x64/src/runtime/asm_amd64.s:1650
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = send wait: context deadline exceeded
        	Test:       	TestIntegrationSelfTracing
    e2e_test.go:369: 
        	Error Trace:	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:369
        	            				D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:205
        	            				C:/hostedtoolcache/windows/go/1.21.12/x64/src/runtime/asm_amd64.s:1650
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = send wait: context deadline exceeded
        	Test:       	TestIntegrationSelfTracing
    e2e_test.go:369: 
        	Error Trace:	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:369
        	            				D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:205
        	            				C:/hostedtoolcache/windows/go/1.21.12/x64/src/runtime/asm_amd64.s:1650
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = send wait: context deadline exceeded
        	Test:       	TestIntegrationSelfTracing
    e2e_test.go:369: 
        	Error Trace:	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:369
        	            				D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:205
        	            				C:/hostedtoolcache/windows/go/1.21.12/x64/src/runtime/asm_amd64.s:1650
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = send wait: context deadline exceeded
        	Test:       	TestIntegrationSelfTracing
    e2e_test.go:369: 
        	Error Trace:	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:369
        	            				D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:205
        	            				C:/hostedtoolcache/windows/go/1.21.12/x64/src/runtime/asm_amd64.s:1650
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = send wait: context deadline exceeded
        	Test:       	TestIntegrationSelfTracing
    e2e_test.go:369: 
        	Error Trace:	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:369
        	            				D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:205
        	            				C:/hostedtoolcache/windows/go/1.21.12/x64/src/runtime/asm_amd64.s:1650
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = send wait: context deadline exceeded
        	Test:       	TestIntegrationSelfTracing
    e2e_test.go:369: 
        	Error Trace:	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:369
        	            				D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:205
        	            				C:/hostedtoolcache/windows/go/1.21.12/x64/src/runtime/asm_amd64.s:1650
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = send wait: context deadline exceeded
        	Test:       	TestIntegrationSelfTracing
    e2e_test.go:369: 
        	Error Trace:	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:369
        	            				D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:205
        	            				C:/hostedtoolcache/windows/go/1.21.12/x64/src/runtime/asm_amd64.s:1650
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = send wait: context deadline exceeded
        	Test:       	TestIntegrationSelfTracing
    e2e_test.go:369: 
        	Error Trace:	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:369
        	            				D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:205
        	            				C:/hostedtoolcache/windows/go/1.21.12/x64/src/runtime/asm_amd64.s:1650
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = send wait: context deadline exceeded
        	Test:       	TestIntegrationSelfTracing
    e2e_test.go:272: 
        	Error Trace:	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:272
        	            				D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:418
        	            				D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:220
        	            				D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/internal/otelarrow/test/e2e_test.go:476
        	Error:      	Not equal: 
        	            	expected: 10000
        	            	actual  : 4664
        	Test:       	TestIntegrationSelfTracing
```

### PR DESCRIPTION
**Pull request by jmacd**
**Description:** Restore a skipped test, after understanding the nature of the problem. 

The problem was mostly addressed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34794, which left the test disabled. The test had been flaky because while testing for an out-of-memory condition, the test could fail for timeout or other reason. To make the test more reliable, this now waits until at least one ArrowTraces span has been received by both components. After one span is available, it checks that the expected log messages are present on both sides.

**Link to tracking Issue:** 
Fixes #34719.

**Testing:** ✅

### CODE REVIEW

**Review by jpkrohling — APPROVED** [review]

**Comment by jmacd** [comment]
Still flaky!

**Comment by jmacd** [comment]
I have been investigating this and found something very curious (and probably buggy) in the integration between grpc-go and the otelgrpc instrumentation used for these tests. In the MemoryLimited test case that is currently skipped, specifically, I seem to have a reproducible case in which streams are started and finished but no spans ever End. I will continue investigating.

**Comment by jmacd** [comment]
PR is green and restores a test. Please merge!

File: internal/otelarrow/test/e2e_test.go [hunk 1]
```diff
@@ -357,12 +367,12 @@ func failureMemoryLimitEnding(t *testing.T, _ testParams, testCon *testConsumer,
 	eSigs, eMsgs := logSigs(testCon.expLogs)
 	rSigs, rMsgs := logSigs(testCon.recvLogs)
 
-	// Test for arrow stream errors.
-	require.Less(t, 0, eSigs["arrow stream error|||code///message///where"], "should have exporter arrow stream errors: %v", eSigs)
+	// Test for arrow Receiver stream errors on both sides.
```
Comment (crobert-1) [inline internal/otelarrow/test/e2e_test.go:370]:
```suggestion
	// Test for arrow receiver stream errors on both sides.
```

**Review by crobert-1 — APPROVED** [review]

**Review by mwear — APPROVED** [review]

**Comment by jmacd** [comment]
(Unrelated test flakes strikes.)

**Comment by crobert-1** [comment]
> (Unrelated test flakes strike.)

I've filed #35033, agreed they're unrelated.

**Comment by jmacd** [comment]
Green!!!

**Review by codeboten — APPROVED** [review]

**Pull request merged**